File: apps/ingestor/src/lib/pipeline.ts
import { classifyReferrer, normalizeProps, parseUrl, type Beacon, type EventRow } from '@seyr/shared';
import { hash64 } from './hash.js';
import { parseUserAgent } from './ua.js';
import { resolveGeo } from './geo.js';
import { clientIp } from './bots.js';

export interface RequestContext {
	headers: Headers;
	userAgent: string;
	salt: string;
	siteId: bigint;
	now?: Date;
}

/**
 * Turn a validated beacon + request into a ClickHouse-ready row. All PII (IP, raw
 * UA) is consumed here to derive the visitor hash, geo, and device — and then
 * dropped; none of it reaches the returned row.
 */
export function buildEventRow(beacon: Beacon, ctx: RequestContext): EventRow {
	const { hostname, pathname, utm_source, utm_medium, utm_campaign } = parseUrl(beacon.u);
	const ip = clientIp(ctx.headers);
	const device = parseUserAgent(ctx.userAgent);
	const geo = resolveGeo(ctx.headers);
	const props = normalizeProps(beacon.m);

	// Daily-rotating, non-durable visitor identity. Domain is included so the same
	// person on two tracked sites never shares a hash.
	const visitorId = hash64(`${ctx.salt}|${ip}|${ctx.userAgent}|${beacon.d}`);

	return {
		site_id: ctx.siteId,
		timestamp: Math.floor((ctx.now?.getTime() ?? Date.now()) / 1000),
		name: beacon.n,
		visitor_id: visitorId,
		// MVP: one session per visitor per day. Sessionization with a 30-min
		// inactivity window is a later refinement (see plan phase 6).
		session_id: visitorId,
		hostname,
		pathname,
		referrer: beacon.r ?? '',
		referrer_source: classifyReferrer(beacon.r ?? '', hostname),
		utm_source,
		utm_medium,
		utm_campaign,
		country_code: geo.country_code,
		region: geo.region,
		city: geo.city,
		browser: device.browser,
		browser_version: device.browser_version,
		os: device.os,
		os_version: device.os_version,
		device: device.device,
		prop_keys: props.keys,
		prop_values: props.values
	};
}

File: apps/ingestor/internal/event/pipeline.go
package event

import (
	"net/http"
	"time"
)

// RequestContext carries the per-request inputs the pipeline derives from.
type RequestContext struct {
	Header    http.Header
	UserAgent string
	Salt      string
	SiteID    uint64
	Now       time.Time
}

// BuildRow turns a validated beacon + request into a ClickHouse-ready row. All
// PII (IP, raw UA) is consumed here to derive the visitor hash, geo, and device,
// then dropped — none of it reaches the returned Row.
func BuildRow(b *Beacon, ctx RequestContext) Row {
	parsed := ParseURL(b.U)
	ip := ClientIP(ctx.Header)
	device := ParseUserAgent(ctx.UserAgent)
	geo := ResolveGeo(ctx.Header)
	keys, values := NormalizeProps(b.M)

	now := ctx.Now
	if now.IsZero() {
		now = time.Now()
	}

	// Daily-rotating, non-durable visitor identity. Domain is included so the
	// same person on two tracked sites never shares a hash.
	visitorID := hash64(ctx.Salt + "|" + ip + "|" + ctx.UserAgent + "|" + b.D)

	return Row{
		SiteID:    ctx.SiteID,
		Timestamp: now.UTC(),
		Name:      b.N,
		VisitorID: visitorID,
		// MVP: one session per visitor per day. A 30-min inactivity window is a
		// later refinement (see plan phase 6).
		SessionID:      visitorID,
		Hostname:       parsed.Hostname,
		Pathname:       parsed.Pathname,
		Referrer:       b.R,
		ReferrerSource: ClassifyReferrer(b.R, parsed.Hostname),
		UTMSource:      parsed.UTMSource,
		UTMMedium:      parsed.UTMMedium,
		UTMCampaign:    parsed.UTMCampaign,
		CountryCode:    geo.CountryCode,
		Region:         geo.Region,
		City:           geo.City,
		Browser:        device.Browser,
		BrowserVersion: device.BrowserVersion,
		OS:             device.OS,
		OSVersion:      device.OSVersion,
		Device:         device.Device,
		PropKeys:       keys,
		PropValues:     values,
	}
}
